lib/improver/lapse_rate.py

```diff
@@ -30,13 +30,87 @@
 # POSSIBILITY OF SUCH DAMAGE.
 """Module containing lapse rate calculation plugins."""
 
-import iris
 import numpy as np
 from numpy.linalg import lstsq
+from scipy.ndimage import generic_filter
+
+import iris
+from iris.analysis.maths import multiply
+from iris.exceptions import CoordinateNotFoundError
+
+from improver.utilities.cube_checker import spatial_coords_match
 from improver.utilities.cube_manipulation import enforce_float32_precision
 from improver.constants import DALR
 
-from scipy.ndimage import generic_filter
+
+def apply_lapse_rate(temperature, lapse_rate, source_orog, dest_orog):
+    """
+    Function to apply a lapse rate adjustment to temperature data forecast
+    at "source_orog" heights, to be applicable at "dest_orog" heights.
+
+    Args:
+        temperature (iris.cube.Cube):
+            Input temperature field to be adjusted
+        lapse_rate (iris.cube.Cube):
+            Cube of pre-calculated lapse rates (units modified in place), which
+            must match the temperature cube
+        source_orog (iris.cube.Cube):
+            2D cube of source orography heights (units modified in place)
+        dest_orog (iris.cube.Cube):
+            2D cube of destination orography heights (units modified in place)
+
+    Returns:
+        adjusted_temperature (iris.cube.Cube):
+            Lapse-rate adjusted temperature field
+    """
+    # check dimensions and coordinates match on input cubes
+    for crd in temperature.coords(dim_coords=True):
+        try:
+            if crd != lapse_rate.coord(crd.name()):
+                raise ValueError(
+                    'Lapse rate cube coordinate "{}" does not match '
+                    'temperature cube coordinate'.format(crd.name()))
+        except CoordinateNotFoundError:
+            raise ValueError('Lapse rate cube has no coordinate '
+                             '"{}"'.format(crd.name()))
+
+    if not spatial_coords_match(temperature, source_orog):
+        raise ValueError(
+            'Source orography projection does not match temperature grid')
```

Comment:
Should this be `x and y coordinates`  rather than `projection`? The projection could be the same for the source orography and temperature grid but coordinates could be different.